File: CSharp/SpecStudioExamplesTestsCSharp/common/AddressTyped.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using System;
using System.Buffers;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using System.Text.Json;
using production;

    public class AddressTyped
    {
        public SimpleText street;
        public SimpleText city;
        public SimpleText state;
        public SimpleText zIP;

        public AddressTyped(SimpleText street, SimpleText city, SimpleText state, SimpleText zIP)
        {
            this.street = street;
            this.city = city;
            this.state = state;
            this.zIP = zIP;
        }

        public AddressString ToAddressString()
        {
            return new AddressString(
                Json.ToText(this.street),
                Json.ToText(this.city),
                Json.ToText(this.state),
                Json.ToText(this.zIP)
            );
        }

        public static List<AddressString> ToStringList(List<AddressTyped> list)
        {
            var result = new List<AddressString>();
            foreach (var t in list) result.Add(t.ToAddressString());
            return result;
        }

        public static List<AddressTyped> FromStringList(List<AddressString> list)
        {
            var result = new List<AddressTyped>();
            foreach (var s in list) result.Add(s.ToAddressTyped());
            return result;
        }

        public void WriteJson(Utf8JsonWriter w)
        {
            w.WriteStartObject();
            w.WriteString("street", Json.ToText(this.street));
            w.WriteString("city", Json.ToText(this.city));
            w.WriteString("state", Json.ToText(this.state));
            w.WriteString("zIP", Json.ToText(this.zIP));
            w.WriteEndObject();
        }

        public string ToJSON()
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer)) { WriteJson(w); }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static AddressTyped FromJsonElement(JsonElement m)
        {
            return new AddressTyped(
                new SimpleText(Json.AsString(Json.Require(m, "street"), "street")),
                new SimpleText(Json.AsString(Json.Require(m, "city"), "city")),
                new SimpleText(Json.AsString(Json.Require(m, "state"), "state")),
                new SimpleText(Json.AsString(Json.Require(m, "zIP"), "zIP"))
            );
        }

        public static AddressTyped FromJSON(string json)
        {
            return FromJsonElement(Json.Parse(json));
        }

        public static string ToJSONList(List<AddressTyped> list)
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer))
            {
                w.WriteStartArray();
                foreach (var item in list) item.WriteJson(w);
                w.WriteEndArray();
            }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static List<AddressTyped> FromJSONList(string json)
        {
            var result = new List<AddressTyped>();
            var root = Json.Parse(json);
            Json.RequireArray(root, "AddressTyped");
            foreach (var e in root.EnumerateArray()) result.Add(FromJsonElement(e));
            return result;
        }

        public override string ToString()
        {
            return $"Street={street}, City={city}, State={state}, ZIP={zIP}";
        }

        public override bool Equals(object? obj)
        {
            if (obj is not AddressTyped other) return false;
            return object.Equals(this.street, other.street)
                && object.Equals(this.city, other.city)
                && object.Equals(this.state, other.state)
                && object.Equals(this.zIP, other.zIP);
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.street);
            h.Add(this.city);
            h.Add(this.state);
            h.Add(this.zIP);
            return h.ToHashCode();
        }
    }
}

File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/AddressTyped.java
package spectable.common;

import java.util.ArrayList;
import java.util.List;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class AddressTyped {
    public SimpleText street;
    public SimpleText city;
    public SimpleText state;
    public SimpleText zIP;

    public AddressTyped(SimpleText street, SimpleText city, SimpleText state, SimpleText zIP) {
        this.street = street;
        this.city = city;
        this.state = state;
        this.zIP = zIP;
    }

    public AddressTyped(AddressString s) {
        this.street = new SimpleText(s.street);
        this.city = new SimpleText(s.city);
        this.state = new SimpleText(s.state);
        this.zIP = new SimpleText(s.zIP);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof AddressTyped)) return false;
        AddressTyped that = (AddressTyped) o;
        return Objects.equals(street, that.street)
            && Objects.equals(city, that.city)
            && Objects.equals(state, that.state)
            && Objects.equals(zIP, that.zIP);
    }

    @Override
    public int hashCode() {
        return Objects.hash(street, city, state, zIP);
    }

    public AddressString toAddressString() {
        return new AddressString(String.valueOf(street), String.valueOf(city), String.valueOf(state), String.valueOf(zIP));
    }

    public static List<AddressTyped> fromStringList(List<AddressString> list) {
        List<AddressTyped> result = new ArrayList<>();
        for (AddressString s : list) result.add(new AddressTyped(s));
        return result;
    }

    public static List<AddressString> toStringList(List<AddressTyped> list) {
        List<AddressString> result = new ArrayList<>();
        for (AddressTyped t : list) result.add(t.toAddressString());
        return result;
    }

    /** Internal plumbing for {@link Json}; use toJSON() for JSON text. */
    public Map<String, Object> toJsonValue() {
        Map<String, Object> m = new LinkedHashMap<String, Object>();
        m.put("street", street == null ? null : street.toString());
        m.put("city", city == null ? null : city.toString());
        m.put("state", state == null ? null : state.toString());
        m.put("zIP", zIP == null ? null : zIP.toString());
        return m;
    }

    public String toJSON() { return Json.write(toJsonValue()); }

    /** Internal plumbing for {@link Json}; use fromJSON(String) for JSON text. */
    public static AddressTyped fromJsonValue(Map<String, Object> m) {
        if (m == null) return null;
        return new AddressTyped(
            new SimpleText(Json.asString(Json.require(m, "street"), "street")),
            new SimpleText(Json.asString(Json.require(m, "city"), "city")),
            new SimpleText(Json.asString(Json.require(m, "state"), "state")),
            new SimpleText(Json.asString(Json.require(m, "zIP"), "zIP")));
    }

    public static AddressTyped fromJSON(String json) {
        return fromJsonValue(Json.parseObject(json));
    }

    public static String toJSONList(List<AddressTyped> list) {
        List<Object> arr = new ArrayList<Object>();
        if (list != null)
            for (AddressTyped item : list) arr.add(item == null ? null : item.toJsonValue());
        return Json.write(arr);
    }

    public static List<AddressTyped> fromJSONList(String json) {
        List<AddressTyped> result = new ArrayList<AddressTyped>();
        for (Object e : Json.parseArray(json))
            result.add(fromJsonValue(Json.asObject(e, "AddressTyped")));
        return result;
    }
}
